### PR TITLE
Windows support for build_test

### PR DESCRIPTION
**Pull request by stdbug**
I have to point out that the approach used in this PR is a bit hacky as it marks `.bat` file as executable on Windows. AFAIK, in order to spawn processes on Windows, Bazel uses `CreateProcessW` function whose [docs](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw) says that caller should use `cmd.exe` as the application name and the batch file as one of `lpCommandLine` arguments. On the other hand, I don't know any Windows flavor that wouldn't run the batch file if it's passed as the application name to `CreateProcessW`

### CODE REVIEW

File: rules/build_test.bzl [hunk 1]
```diff
@@ -16,16 +16,44 @@
...
+    return [DefaultInfo(
+        files = depset([executable]),
+        executable = executable,
+        runfiles = ctx.runfiles(files = runfiles),
+    )]
+
+_empty_test = rule(
+    implementation = _empty_test_impl,
+    attrs = {
+        "data": attr.label_list(),
```
Comment (FaBrand) [inline rules/build_test.bzl:45]:
```suggestion
        "data": attr.label_list(allow_files = True),
```
You can use the `allow_files` to make the code above simpler. No need to [call to_list() in a loop](https://docs.bazel.build/versions/master/skylark/depsets.html#performance)

File: rules/build_test.bzl [hunk 2]
```diff
@@ -16,16 +16,44 @@
...
+    )
+
+    runfiles = []
+    for entry in ctx.attr.data:
+        runfiles += entry.files.to_list()
+
+    return [DefaultInfo(
+        files = depset([executable]),
+        executable = executable,
+        runfiles = ctx.runfiles(files = runfiles),
```
Comment (FaBrand) [inline rules/build_test.bzl:39]:
```suggestion
        runfiles = ctx.runfiles(files = ctx.files.data),
```

File: rules/build_test.bzl [hunk 3]
```diff
@@ -16,16 +16,44 @@
...
+    ctx.actions.write(
+        output = executable,
+        is_executable = True,
+        content = "exit 0",
+    )
+
+    runfiles = []
+    for entry in ctx.attr.data:
+        runfiles += entry.files.to_list()
+
```
Comment (FaBrand) [inline rules/build_test.bzl:35]:
```suggestion
```
Not necessary with `ctx.files.data`

**Review by FaBrand — COMMENTED** [review]
Some suggestions to make the empty test more concise with its runfiles

File: rules/build_test.bzl [hunk 4]
```diff
@@ -16,16 +16,40 @@
 
 load("//lib:new_sets.bzl", "sets")
 
+def _empty_test_impl(ctx):
+    is_windows = select({
+        "@bazel_tools//src/conditions:host_windows": True,
+        "//conditions:default": False,
+    })
+    extension = ".bat" if is_windows else ".sh"
```
Comment (FaBrand) [inline rules/build_test.bzl:24]:
```suggestion
    extension = ".bat" if ctx.attr.is_windows else ".sh"
```

File: rules/build_test.bzl [hunk 5]
```diff
@@ -16,16 +16,40 @@
...
+    return [DefaultInfo(
+        files = depset([executable]),
+        executable = executable,
+        runfiles = ctx.runfiles(files = ctx.files.data),
+    )]
+
+_empty_test = rule(
+    implementation = _empty_test_impl,
+    attrs = {
+        "data": attr.label_list(allow_files = True),
```
Comment (FaBrand) [inline rules/build_test.bzl:41]:
```suggestion
        "data": attr.label_list(allow_files = True),
        "is_windows": attr.bool(mandatory = True),
```

File: rules/build_test.bzl [hunk 6]
```diff
@@ -75,15 +99,13 @@ def build_test(name, targets, **kwargs):
...
             **genrule_args
         )
 
-    native.sh_test(
+    _empty_test(
         name = name,
-        # TODO: Does this need something else for Windows?
-        srcs = ["@bazel_skylib//rules:empty_test.sh"],
         data = test_data,
         size = kwargs.pop("size", "small"),  # Default to small for test size
```
Comment (FaBrand) [inline rules/build_test.bzl:110]:
```suggestion
        size = kwargs.pop("size", "small"),  # Default to small for test size
        is_windows = select({
                "@bazel_tools//src/conditions:host_windows": True,
                "//conditions:default": False,
        }),
```

File: rules/build_test.bzl (hunk 4 above)
Comment (FaBrand) [reply rules/build_test.bzl:24]: This is a bug, the select is not evaluated as desired in this context.
However it is truthy, therfore the if does parse. Therefore the condition always evaluates to `True` and extension .bat is always used.

File: rules/build_test.bzl (hunk 6 above)
Comment (FaBrand) [reply rules/build_test.bzl:110]: This is evaluatited in the `BUILD`, file, therefore the select is resolved as expected.

**Review by meteorcloudy — APPROVED** [review]

**Review by tetromino — APPROVED** [review]

**Pull request merged**